lib/instrumentation/modules/tedious.js

```diff
@@ -74,9 +74,17 @@ module.exports = function (tedious, agent, { version, enabled }) {
           return super.makeRequest(...arguments)
         }
 
-        const preparing = request.sqlTextOrProcedure === 'sp_prepare'
-        const params = request.parametersByName
-        const sql = (params.statement || params.stmt || {}).value
+        let sql
+        let preparing
+        if (semver.satisfies(version, '^1.9.0 || 2.x || 3.x || ^4.0.1 || 5.x || 6.x || 7.x || 8.x || 9.x || 10.x')) {
+          preparing = request.sqlTextOrProcedure === 'sp_prepare'
+          const params = request.parametersByName
+          sql = (params.statement || params.stmt || {}).value
+        } else {
+          preparing = payload.procedure === 'sp_prepare'
+          const params = { stmt: payload.parameters.find(({ name }) => name === 'stmt'), statement: payload.parameters.find(({ name }) => name === 'statement') }
+          sql = (params.statement || params.stmt || {}).value || request.sqlTextOrProcedure
+        }
```

Comment:
```suggestion
        let sql
        let preparing
        if (payload.parameters !== undefined) {
          // This looks like a tedious `RpcRequestPayload` instance for tedious
          // version >=v11.0.10, when RPC parameter handling was refactored
          // (https://github.com/tediousjs/tedious/pull/1275).
          preparing = payload.procedure === 'sp_prepare'
          const stmtParam = (payload.parameters.find(({ name }) => name === 'statement') ||
            payload.parameters.find(({ name }) => name === 'stmt'))
          sql = stmtParam ? stmtParam.value : request.sqlTextOrProcedure
        } else {
          preparing = request.sqlTextOrProcedure === 'sp_prepare'
          const params = request.parametersByName
          sql = (params.statement || params.stmt || {}).value
        }
```


I dug in a little bit to why this change to use `payload`. It was due to this change: https://github.com/tediousjs/tedious/pull/1275
This was released in tedious@11.0.10, so the `semver.satisfies` isn't quite correct. Using, e.g., tedious@11.0.0 will crash in the tests:

```
% node test/instrumentation/modules/tedious.test.js
TAP version 13
# prepare / execute
(node:19416) [DEP0123] DeprecationWarning: Setting the TLS ServerName to an IP address is not permitted by RFC 6066. This will be ignored in a future version.
(Use `node --trace-deprecation ...` to show where the warning was created)
/Users/trentm/el/apm-agent-nodejs17/lib/instrumentation/modules/tedious.js:85
          const params = { stmt: payload.parameters.find(({ name }) => name === 'stmt'), statement: payload.parameters.find(({ name }) => name === 'statement') }
                                                    ^

TypeError: Cannot read properties of undefined (reading 'find')
    at Connection.makeRequest (/Users/trentm/el/apm-agent-nodejs17/lib/instrumentation/modules/tedious.js:85:53)
    at Connection.prepare (/Users/trentm/el/apm-agent-nodejs17/node_modules/tedious/lib/connection.js:2097:10)
    at /Users/trentm/el/apm-agent-nodejs17/test/instrumentation/modules/tedious.test.js:136:16
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

Another thing we noted in our `hapi` instrumentation is that calling `semver.satisfies()` can be quite slow, so I've used feature sniffing (checking for `payload.parameters`) rather than version checking, in my stab at this.  Does my attempt look good to you?  I haven't yet checked which with all versions.

```
TAV=tedious ./node_modules/.bin/tav
```

Reply:
Great suggestion, this is much better approach. 
I had worries about performance, but as I saw it used heavily in repository.

Reply:
Is there need to change also [on line 69](https://github.com/elastic/apm-agent-nodejs/pull/2705/files#diff-abd6d394723d7638c6714a6d0a68ba4c96f030125aed364ea7926f386ad046f2R69)?

Reply:
I don't think that guard on line 69 *needs* to change. I think it requires a bit of digging to understand what this `makeRequest` is able to handle. IIUC, it is written to handle just calls where `_packageType === TYPE.RPC_REQUEST`, but that `TYPE` isn't exposed by the tedious library because we are deep in its private interfaces here. :)

However, clarifying all that need not happen as part of this change. The current guard suffices, and what you have added is a definite improvement.